### ISSUE TITLE
Tagged-union (literal-attribute) narrowing doesn't distribute over a prior AlwaysFalsy intersection

### ISSUE THREAD

**Issue by ainsleymcgrath** [astral-sh/ty#4100]
Tagged-union narrowing on a `Literal`-typed attribute (#1479) works correctly on its own, including for inherited fields:

```py
from typing import Literal
from pydantic import BaseModel

class AssetReference(BaseModel):
    variant: Literal["persisted"]
    uuid: str

class Persisted(AssetReference):
    pass

class Synthetic(BaseModel):
    variant: Literal["synthetic"]
    label: str

def f(model: Persisted | Synthetic) -> None:
    if model.variant == "persisted":
        reveal_type(model)  # revealed: Persisted -- correct
        print(model.uuid)
    else:
        reveal_type(model)  # revealed: Synthetic -- correct
        print(model.label)
```

But once an earlier truthiness/`Optional` check has narrowed the union to `X & ~AlwaysFalsy` members, the subsequent tag-equality narrow no longer distributes across that intersection — both branches still see the full (un-narrowed) union:

```py
from typing import Literal
from pydantic import BaseModel

class AssetReference(BaseModel):
    variant: Literal["persisted"]
    uuid: str

class Persisted(AssetReference):
    pass

class Synthetic(BaseModel):
    variant: Literal["synthetic"]
    label: str

class Request(BaseModel):
    set_packaging_model: Persisted | Synthetic | None

class Resolver:
    def __init__(self, request: Request) -> None:
        self._request = request

    def resolve(self) -> None:
        if not self._request.set_packaging_model:
            return None
        model = self._request.set_packaging_model
        if model.variant == "persisted":
            reveal_type(model)  # revealed: (Persisted & ~AlwaysFalsy) | (Synthetic & ~AlwaysFalsy)
            print(model.uuid)   # error[unresolved-attribute]: `uuid` not defined on `Synthetic & ~AlwaysFalsy`
        else:
            reveal_type(model)  # revealed: (Persisted & ~AlwaysFalsy) | (Synthetic & ~AlwaysFalsy)
            print(model.label)  # error[unresolved-attribute]: `label` not defined on `Persisted & ~AlwaysFalsy`
```

Expected: the tag-equality narrow should distribute over each union member's `& ~AlwaysFalsy` intersection the same way it does for a bare union, collapsing to `Persisted & ~AlwaysFalsy` / `Synthetic & ~AlwaysFalsy` respectively.

`ty 0.0.64 (5e64a131b 2026-07-27)`.

Found while investigating a real diff (Pydantic discriminated-union pattern, `variant: Literal[...]` discriminant on a hierarchy of `BaseModel`s with an `Optional[...]` outer field) where mypy passes and `ty` produces 22 false-positive `unresolved-attribute` diagnostics.

**Comment by carljm** [comment]
Thanks! This is another one that would be trivially a non-issue with #2233, but it also looks fairly easy to handle these intersections in the tagged-union narrowing code.
